tools/generator/dfg/sam/sam_identifier.py

```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013     , Kevin Läufer
+# Copyright (c) 2013-2014, Niklas Hauser
+# Copyright (c)      2016, Fabian Greif
+# All rights reserved.
+
+import re
+import logging
+
+from modm_devices.device_identifier import DeviceIdentifier
+
+LOGGER = logging.getLogger("dfg.sam.identifier")
+
+class SAMIdentifier:
+    """ SAMIdentifier
+    A class to parse SAM device strings, e.g. ATSAMD21E15A-MUT.
+    Device names are organized as follows:
+       ATSAMD     21         E        15        A     -    M          U           T
+      {family} {Series} {Pin Count} {Flash} {Revision} {pacakge} {Temp Grade} {carrier}
+    """
+
+    @staticmethod
+    def from_string(string):
+        string = string.lower()
+
+        # SAM platform with SAMD, SAML, SAMC, SAM4, SAMG, SAMS, SAME, and SAMV
+        if string.lower().startswith("atsam"):
+            # revision is the silicon revision of the chip
+            matchString = r"ATSAM(?P<family>[A-Z])(?P<name>[0-9]{2})(?P<pin>.)(?P<size>[0-9]{2})(?P<revision>.+)\."
+            match = re.search(matchString, string)
+            if match:
+                i = DeviceIdentifier()
```

Comment:
You also need to set the format string to `{platform}{family}{series}{pin}{flash}{revision}-{package}`, otherwise the identifier cannot format itself. This string is also used to reverse match the input string without having to define a parser for it.

See also: https://blog.salkinium.com/modm-devices/#device-identifiers